### PR TITLE
Item movement test harness + fixes for 8 movement bugs (#10253)

### LINKED ISSUE #3573: Transfer search can fail to transfer all items when an intermediate transfer needs to occur.
If you're transferring search results to the vault, and one of the matches items is equipped on a character, and there is nothing else on that character to swap it for, DIM will transfer and item from the vault in order to unequip the equipped item. The problem is is that sometimes DIM will transfer an item that still matches the search results, resulting in an incomplete transfer.

This might be confusing, so i made this video demonstrating the bug https://gfycat.com/VeneratedHopefulAtlasmoth You'll not that once the transfer is complete, there are still items on my character that match the search.

The solution, in my mind, is when DIM pulls something from the vault for the swap, it should prefer items that are not in the current 'loadout' being transferred.

### LINKED ISSUE #7935: strange coin postmaster infinite loop
### DIM Version

7.5.1, non-beta/orange

### Browser Details

firefox 97.0.1 64 bit

### OS Details

windows 10 pro, 19044.1526

### Describe the bug

Seems to be an infinite loop when trying to pull strange coins from postmaster on a character other than one i am currently on (with specific conditions described below). Works fine to pull from postmaster on character that am currently on.

Vault at 499 and character consumables full. It starts putting a currency in vault then pulling another out of vault to character and goes thru some list (it did start repeating). If vault is at 500 and there is an empty space on character it starts the same loop just the opposite direction first.

I canceled the action and moved a piece of armor from vault to character (so 2 open spaces between char and vault) and then when trying to move strange coin from different postmaster it works. Testing more, if char has 1 open and vault has 1 open, it just moves like you would expect. If char has 2 open and vault is at 500, it moves a currency to have 1 open in each then pulls the coin(s). If you have 2 open in vault it moved a currency to where both have 1 open slot then pulls like expected.

In the case where you have a non-full stack of strange coins on character, requires 1 empty slot in vault and will move a currency from vault to character to meet that condition (character can be full since the non-full stack counts as a "space", but vault needs 1 space).

Seems like the logic is trying to go thru the other character (owner of postmaster) to then go to vault and to current character (like pulling armor/weapons from other characters postmaster works). As soon as coins hit inventory, it does stops, but it tries to clear too many spaces beforehand.

### Logs

```shell
TopSitesFeed: Failed to fetch data from Contile server: NetworkError when attempting to fetch resource. TopSitesFeed.jsm:207
TopSitesFeed: Contile endpoint returned unexpected status: 503 TopSitesFeed.jsm:181
TopSitesFeed: Failed to fetch data from Contile server: JSON.parse: unexpected end of data at line 1 column 1 of the JSON data TopSitesFeed.jsm:207
Content Security Policy: Ignoring “'unsafe-inline'” within script-src or style-src: nonce-source or hash-source specified
Content Security Policy: Ignoring “'unsafe-inline'” within script-src or style-src: nonce-source or hash-source specified 2
Content Security Policy: Ignoring “'unsafe-inline'” within script-src: ‘strict-dynamic’ specified
Content Security Policy: Ignoring “https:” within script-src: ‘strict-dynamic’ specified
Content Security Policy: Ignoring “http:” within script-src: ‘strict-dynamic’ specified
sendRemoveListener on closed conduit {7a7a4a92-a2a0-41d1-9fd7-1e92480d612d}.52913997086723 ConduitsChild.jsm:112
    _send resource://gre/modules/ConduitsChild.jsm:112
    _send self-hosted:1282
    removeListener resource://gre/modules/ExtensionChild.jsm:670
    removeListener resource://gre/modules/ExtensionChild.jsm:905
    onChanged chrome://extensions/content/child/ext-storage.js:332
    removeListener resource://gre/modules/ExtensionCommon.jsm:2511
    revoke resource://gre/modules/ExtensionCommon.jsm:2533
    close resource://gre/modules/ExtensionCommon.jsm:2538
    unload resource://gre/modules/ExtensionCommon.jsm:899
    close resource://gre/modules/ExtensionContent.jsm:935
    destroyed resource://gre/modules/ExtensionContent.jsm:1011
    observe resource://gre/modules/ExtensionContent.jsm:1029
Promise rejected after context unloaded: Actor 'Conduits' destroyed before query 'RuntimeMessage' was resolved
index.js:4937
    sendMessage moz-extension://ce12f7d3-0eee-4b4e-84cd-85a597e84060/inject/index.js:4937
e is undefined h0-main.js:1
    d moz-extension://2471111b-0fae-4032-9d73-d6ab2edfe90d/h0-main.js:1
    apply self-hosted:2732
    applySafeWithoutClone resource://gre/modules/ExtensionCommon.jsm:635
    wrapPromise resource://gre/modules/ExtensionCommon.jsm:837
    withLastError resource://gre/modules/ExtensionCommon.jsm:756
    wrapPromise resource://gre/modules/ExtensionCommon.jsm:825
Unchecked lastError value: Error: Invalid window ID: 7427 h0-main.js:1
    d moz-extension://2471111b-0fae-4032-9d73-d6ab2edfe90d/h0-main.js:1
Content Security Policy: Ignoring “'unsafe-inline'” within script-src or style-src: nonce-source or hash-source specified
```

### LINKED ISSUE #8418: Item move shouldn't consider items in the exclusion list for equipping to replace an equipped item
If you're bulk moving a bunch of stuff, and one of the items is equipped, the move algo shouldn't ever select one of the items being moved as a replacement for the equipped item. Need to make sure the exclusion list is properly populated and that it's consulted.

### LINKED ISSUE #8506: Item Move Fails with Specific Search
### DIM Version

Version 7.22.0.1001823 (beta), built on 6/19/2022, 8:01:45 PM

### Browser Details

102.0.5005.115 (Official Build) (64-bit) 

### OS Details

openSUSE Leap 15.4

### Describe the bug

When I use the search:

`is:movable -is:armor -is:weapon -is:ships -is:vehicle -is:ghost -name:"sweaty confetti" -name:"paraversal haul" -is:transmat -is:invault`

And then attempt to transfer items to the vault, I get a message that says the move was successful, but the items aren't moved.  But with the search:

`is:movable -is:armor -is:weapon -is:ships -is:vehicle -is:ghost -name:"sweaty confetti" -name:"paraversal haul" -is:transmat`

The items that are not in the vault move.

### Logs

```shell
[loadout] Apply loadout Filtered Items to Vault
log.ts:58 Loadout Application: 53.009033203125 ms
```

### LINKED ISSUE #8872: Hymn of Desecration breaks moves using a filtered search
### DIM Version

Version 7.47.0.1002112 (beta), built on 12/17/2022, 11:11:22 AM

### Browser Details

Chrome 108.0.5359.125 (Official Build) (64-bit)

### OS Details

Windows 10

### Describe the bug

1) I created a filtered search using "is:consumables is:movable" and then attempted to move all of the items found into my vault.
2) The move status popup appeared and the items began to move into the vault normally.
3) Once the move reached the Hymn of Desecration, the move paused. DIM then began filling my character inventories, removing weapons and armor from the vault.
4) Once the character inventory was full on all three of my characters, the move completed, save for the Hymn of Desecration, producing a '1 item error: there's not enough room to move this item' error.

Attempting the opposite, moving consumables from the vault into the character, moves all of the consumables normally, save for the Hymn, which fails with the same error message about not having enough room to move the item. Items are not removed from the characters and pushed into the vault when this happens.

If I move the Hymn of Desecration on its own, it moves normally and does not produce any error messages. This issue appears to be specific to a filtered search, as filtering to just the Hymn of Desecration will produce the same issue, while drag/drop or using the vault buttons on the item appear to function normally.

### Logs

_No response_

### LINKED ISSUE #9121: Can't directly equip an exotic if we need to pull from vault to get a replacement
Apparently if you:

1. Have only an exotic equipped in one armor slot (no held, unequipped items)
2. Double-click an exotic armor that is both in the vault and for a different slot than the exotic in the first step

DIM will not pull a piece from the vault to swap in for the first exotic, and the second exotic will fail to equip.

### LINKED ISSUE #9416: Loadout apply exclusion bugs
I have one reproducer for at least some of these, but there may be other problems.

These items match `is:heavy -inloadout:mgloadout`:

grafik

I bulk transfer to the vault, the result is this:

grafik

This is clearly wrong -- an item that was matched by the search (and should thus stay in the vault) was picked as a replacement item. The correct behavior is choosing one of the two unmatched machine guns in the vault as a replacement.

I identified a number of issues in loadout application:

1. Bulk dequipping and single dequipping use different exclusion logic:
https://github.com/DestinyItemManager/DIM/blob/8df972cb5a4a067126db842e4a1fd78f561c89e2/src/app/loadout-drawer/loadout-apply.ts#L397-L417 Here `involvedItems` does not ensure vaulted but matched items are in the `moveSession`. They are in `applicableLoadoutItems`, however, so non-bulk dequipping is weaker and fails to include the vaulted items that shouldn't move.
2. Single movement does not convert move session items to exclusions: https://github.com/DestinyItemManager/DIM/blob/8df972cb5a4a067126db842e4a1fd78f561c89e2/src/app/inventory/item-move-service.ts#L340-L356 Here the `MoveSession` could contain items we shouldn't choose as a replacement, but no exclusion list is passed to `getSimilarItem`.
3. Not reproducable with the above, but `Exclusion`s are matched by ID and hash https://github.com/DestinyItemManager/DIM/blob/8df972cb5a4a067126db842e4a1fd78f561c89e2/src/app/inventory/item-move-service.ts#L1172-L1173 but `exclusions` are derived from `applicableLoadoutItems` https://github.com/DestinyItemManager/DIM/blob/8df972cb5a4a067126db842e4a1fd78f561c89e2/src/app/loadout-drawer/loadout-apply.ts#L412-L415 which in turn is just a sorted list of raw loadout items before resolution https://github.com/DestinyItemManager/DIM/blob/8df972cb5a4a067126db842e4a1fd78f561c89e2/src/app/loadout-drawer/loadout-apply.ts#L244-L248, so IDs are unreliable since we have resolution behavior for shaped items where IDs change. Haven't come up with a reproducer for that yet.

### LINKED ISSUE #10046: Multiple moves overlap badly?
If you request a move of the same item twice (say, when b.net is slow), the second one fails. It should finish with a noop.

### LINKED ISSUE #10253: Testing: Item Movement
There are a number of bugs or at least weird corner cases with how item movement works. I'd love for us to fix them, but the item movement logic is very complex and the chances of any fix breaking something else is high. 

To make fixes possible, we should have a robust test suite for item movement. That will allow us to change things without fear of breaking existing behavior, and allow us to reproduce bugs in a test environment and be sure they're fixed. 

There are a few steps here:

1. Create an inventory test fixture that lets us easily configure different scenarios. This can start with my profile export but then create conditions like varying levels of free space in different buckets, different stacks of consumables, etc. It should be simple to set up states that allow us to reproduce bugs. 
2. Set up some sort of mocking for the Bungie.net transfer/lock APIs so we can simulate different outcomes. 
3. Build out a suite of tests for existing item movement scenarios, guided by code coverage and comments to try and capture all the edge cases that are currently handled. 
4. Write tests that reproduce our open bugs and then fix them. See the "Feature: Item Movement" label. 

I would love help with this, as I don't feel like I have time to really make progress on it.

### LINKED ISSUE #11648: Destiny 1 - Max Light does not move Ghost Shells
### DIM Version

Version 8.102.0 (release)

### Browser Details

DIM Appstore 7 - iOS 18.6.2

### OS Details

iOS

### Describe the bug

When clicking a character and selecting Max light, it does not give them the max light ghost shell.

### Logs

```shell

```

### PR DESCRIPTION
**Pull request by delphiactual**
Builds the item-movement test infrastructure asked for in #10253, then uses it to reproduce and fix a batch of open "Feature: Item Movement" bugs. Every fix is TDD'd - the regression test was verified to fail before the fix and pass after.

  ## Test infrastructure (#10253)

  - `src/testing/move-item-test-utils.ts`: seeds the real Redux store from the sample profile and drives the actual move logic (`executeMoveItem`, `moveItemTo`, and full `applyLoadout` runs). Includes fixture helpers for setting up edge cases (full buckets, partial/unique stacks, postmaster items, cross-store replacements).
  - The Bungie.net write APIs are mocked, so moves never hit the network; the in-memory model is driven by the reducer.
  - 52 tests across `item-move-service`, `item-move-matrix`, `loadout-apply`, `auto-loadouts`, and `move-item`, including a parameterized movement matrix (source/destination pairs, destination fullness, stack merge/overflow, postmaster pulls), and covering transfers (all directions), move-asides, stacking, equip/de-equip, the one-exotic rule, error paths, and end-to-end loadout application.

  ## Bugs fixed

  | Issue | Fix |
  |-------|-----|
  | #8872, #8506 | Unique-stack consumables (Hymn of Desecration, Ghost Fragments) failed to move via a filtered search. loadout-apply only marked equips/dequips as "involved", so the consumables space penalty made the destination look full → silent skip / move-aside cascade. Now marks every moved item involved. |
  | #9121 | Equipping a vault exotic for a different slot didn't de-equip the existing exotic — the blind-move fast path skipped the one-exotic check. Skip the blind move when equipping an item with an equipping label. |
  | #8418 / #9416 pt2 | A de-equip could pick an item the user was actively moving as the replacement. The move session now carries its involved items as exclusions, passed to `getSimilarItem`. |
  | #3573 / #9416 pt1 | Bulk search-to-vault could pull a search-matched item out of the vault as a de-equip replacement. The session is now built from all resolved loadout items, not just the ones that need to move. |
  | #9416 pt3 | `equipItems` bulk-equipped items assuming they were already in the character's inventory; de-equip replacements pulled from the vault silently failed. `loadout-apply` now moves those replacements onto the target store before bulk-equipping, and `equipItems` documents the on-store precondition. |
  | #9416 pt4 | Exclusions matched by `{id, hash}`, but crafted/shaped items resolve to a new id (by `craftedDate`), so a loadout's own crafted item wasn't excluded. Now passes resolved items as exclusions. (No prior repro existed - this PR adds one.) |
  | #11648 | D1 "Max Light" left out the Ghost Shell / Artifact. In D1 these contribute to Light but aren't weapons/armor; they're now included for D1 stores. |
  | #10046 | Requesting the same move twice (e.g. clicking again while Bungie.net is slow) failed the second one on a stale item reference. The queued move now re-resolves the live item and no-ops if it's already at the destination. |

  ## Investigated, no change needed

  - #6895 - same-hash unique-stack char-to-char scenario already stops cleanly (a `StackFull` error, no runaway). The 2021 "transfer everything" cascade relied on a cross-hash game-level constraint DIM doesn't model and isn't reproducible today. Added a regression test to lock in the clean-stop behavior; left the "smart swap" enhancement out as out-of-scope.
  
  
Closes #8872
Closes #8506
Closes #9121
Closes #8418
Closes #3573
Closes #9416
Closes #11648
Closes #10046
Closes #7935
Closes #10253

### CODE REVIEW

File: src/testing/move-item-test-utils.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+ * returns stale data. Fixture helpers must use this instead.
+ */
+export function itemsInBucketUncached(store: DimStore, bucketHash: number): DimItem[] {
+  return store.items.filter((i) => i.location.hash === bucketHash && !i.location.inPostmaster);
+}
+
+/**
+ * Build a *fresh* set of DimStores from the sample profile. Unlike
+ * `getTestStores` in test-utils, this is not memoized, so each call returns new
+ * mutable store/item objects that fixture helpers are free to modify.
```
Comment (bhollis) [inline src/testing/move-item-test-utils.ts]: Same question - why would we mutate? The cache should still be valid if we update items immutably.

We can use immer in the helpers.

File: src/testing/move-item-test-utils.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+ */
+
+const testAccount = rawTestAccount as unknown as DestinyAccount;
+const thunkDispatch = reduxStore.dispatch as DimThunkDispatch;
+
+/**
+ * Non-memoized version of findItemsByBucket. `findItemsByBucket` is weakMemoized
+ * on the store object, so reading it after mutating `store.items` in place
+ * returns stale data. Fixture helpers must use this instead.
+ */
```
Comment (bhollis) [inline src/testing/move-item-test-utils.ts:50]: Why do we need to mutate `store.items` instead of changing them immutably through the reducers?

File: src/app/loadout-drawer/loadout-apply.ts [hunk 3]
```diff
@@ -448,18 +459,16 @@ function doApplyLoadout(
...
           );
           try {
             const result = await dispatch(
-              equipItems(
-                getStore(getStores(), owner)!,
-                itemsToEquip,
-                applicableLoadoutItems,
-                moveSession,
-              ),
+              equipItems(getStore(getStores(), owner)!, itemsToEquip, involvedItems, moveSession),
```
Comment (bhollis) [inline src/app/loadout-drawer/loadout-apply.ts]: I feel like we missed a prettier invocation here?

File: src/app/inventory/item-move-service.ts [hunk 4]
```diff
@@ -76,6 +76,13 @@ export interface MoveSession {
    * Contains instanceIds, or for uninstanced items, item hashes.
    */
   involvedItems: Set<string | number>;
+  /**
+   * The explicitly-involved items, as exclusions. When we need to de-equip an
+   * item to move it, we must not choose one of these as the replacement to
+   * equip - they're items the user is actively moving, so equipping one would
+   * fight the move it's part of.
+   */
+  involvedItemExclusions: Exclusion[];
```
Comment (bhollis) [inline src/app/inventory/item-move-service.ts]: It feels like this doesn't need to be separate from involvedItems and/or involvedItems could just carry both id and hash?

File: src/app/inventory/item-move-service.ts [hunk 5]
```diff
@@ -268,6 +278,20 @@ export function equipItems(
...
+    // You can only equip items that are in the character's inventory. The items
+    // we're handed (e.g. replacements chosen by getSimilarItem) may live in the
+    // vault or on another character, so move each one onto the store first. See
+    // #9416 (point 3). Items already on the store are left as-is.
+    const itemsOnStore: DimItem[] = [];
+    for (const i of items) {
+      itemsOnStore.push(
+        i.owner === store.id
+          ? i
+          : await dispatch(executeMoveItem(i, store, { equip: false }, session)),
```
Comment (bhollis) [inline src/app/inventory/item-move-service.ts]: loadout-apply already moves items into the right place in the common case that this is called - an unstated prerequisite of equipItems is that the items should be in the desired store. I can see where we missed that in the dequip loop but it might be better to either explicitly move the dequip items (from getSimilarItem) onto the right store first.

File: src/app/inventory/item-move-service.test.ts [hunk 6]
```diff
@@ -0,0 +1,825 @@
...
+        !i.uniqueStack &&
+        !i.notransfer &&
+        !i.location.inPostmaster &&
+        vault.items.some((v) => v.hash === i.hash && !v.location.inPostmaster),
+    );
+    if (item) {
+      return { item, source };
+    }
+  }
+  return undefined;
```
Comment (bhollis) [inline src/app/inventory/item-move-service.test.ts]: This should probably be an assert failure? Same for the other find helpers

File: src/app/inventory/item-move-service.test.ts [hunk 7]
```diff
@@ -0,0 +1,825 @@
...
+    transferMock.mockClear();
+    equipMock.mockClear();
+    transferMock.mockResolvedValue({});
+    equipMock.mockResolvedValue({});
+  });
+
+  it('moves an item from the vault to a character', async () => {
+    const stores = await buildFreshStores();
+    const vault = getVault(stores)!;
+    const character = stores.find((s) => !s.isVault)!;
```
Comment (bhollis) [inline src/app/inventory/item-move-service.test.ts]: Any reason this wouldn't be in `beforeEach`?

File: src/app/inventory/item-move-service.test.ts [hunk 8]
```diff
@@ -0,0 +1,825 @@
...
+
+    const item = findTransferableWeapon(vault)!;
+    expect(item).toBeDefined();
+
+    // Fill the destination bucket on the character, and make every item in it
+    // un-moveable so DIM can't free up a slot.
+    setBucketFreeSlots(character, item.bucket.hash, 0);
+    for (const i of itemsInBucketUncached(character, item.bucket.hash)) {
+      i.notransfer = true;
+    }
```
Comment (bhollis) [inline src/app/inventory/item-move-service.test.ts]: Maybe a better (or different?) test would be to fill up every bucket on every character + vault. Actually probably more tests - one where all the buckets matching the type are full, but there's space in the vault, one where the vault is full and the buckets are full but other characters/itemtypes have space, and one where everything is totally full.

File: src/app/inventory/item-move-service.test.ts [hunk 9]
```diff
@@ -0,0 +1,825 @@
...
+
+    const newStores = getStores();
+    const newSource = newStores.find((s) => s.id === source.id)!;
+    // All of it left the source; the vault's total reflects the combined amount
+    // (regardless of how many physical stacks it ends up in).
+    expect(amountOfItem(newSource, item)).toBe(0);
+    expect(amountOfItem(getVault(newStores)!, item)).toBe(startVault + startSource);
+  });
+
+  it('pulls a lost item out of the postmaster onto its character', async () => {
```
Comment (bhollis) [inline src/app/inventory/item-move-service.test.ts]: We should repeat the postmaster tests for all of the "bucket full", "vault full", etc. combos. Maybe worth making it a parameterized/table test helper.

File: src/app/inventory/item-move-service.test.ts [hunk 10]
```diff
@@ -0,0 +1,825 @@
...
+
+    expect(transferMock).toHaveBeenCalled();
+    // It left the postmaster and landed in its real destination bucket.
+    expect(moved.location.inPostmaster).toBeFalsy();
+    expect(moved.location.hash).toBe(destinationBucket);
+    const newOwner = getStores().find((s) => s.id === owner.id)!;
+    expect(newOwner.items.some((i) => i.id === item.id && !i.location.inPostmaster)).toBe(true);
+  });
+
+  it('pulls a lost item out of the postmaster onto another character', async () => {
```
Comment (bhollis) [inline src/app/inventory/item-move-service.test.ts]: For all of the move scenarios we should also make sure to test stacked items vs. not stacked. Building up a kind of test matrix.

File: src/app/inventory/item-move-service.test.ts [hunk 11]
```diff
@@ -0,0 +1,825 @@
...
+    const { move } = setupMoveTestStore(stores);
+
+    await expect(move(extra, character, { amount: 1 })).rejects.toThrow(DimError);
+    expect(transferMock).not.toHaveBeenCalled();
+  });
+
+  // Regression tests for issues #8872 / #8506: bulk-moving consumables via a
+  // filtered search (which goes through loadout-apply) broke on unique-stack
+  // consumables like Hymn of Desecration and Ghost Fragments.
+  it('moves a large unique-stack consumable to the vault in one transfer (#8872)', async () => {
```
Comment (bhollis) [inline src/app/inventory/item-move-service.test.ts:265]: The stack tests should cover moving a stack to somewhere that doesn't have any of that stack, a place that has one stack but room for the incoming stack, a place that has one stack but will overflow to a second item, a place where the destination stack can't overflow because it has to be a single item, etc.

File: src/app/inventory/item-move-service.test.ts [hunk 12]
```diff
@@ -0,0 +1,825 @@
...
+    // single clean move (and may ultimately fail). This is why loadout-apply
+    // must mark moved items as involved.
+    await move(item, vault, { involvedItems: [] }).catch(() => undefined);
+    expect(transferMock.mock.calls.length).toBeGreaterThan(1);
+  });
+
+  // Regression test for issue #9121: equipping an exotic pulled from the vault
+  // must still de-equip the existing exotic, even when the replacement for that
+  // exotic's slot has to come from the vault too.
+  it('de-equips the existing exotic when equipping a vault exotic that needs a vault replacement (#9121)', async () => {
```
Comment (bhollis) [inline src/app/inventory/item-move-service.test.ts:296]: Yeah these test cases are all good but it's making the case for an auto-generated set of "all unique movement possibilities".

File: src/app/loadout-drawer/auto-loadouts.test.ts [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+} as unknown as DimStore;
+
+const d2Store = {
+  id: 'char1',
+  destinyVersion: 2,
+  classType: DestinyClass.Titan,
+  isVault: false,
+} as unknown as DimStore;
+
+describe('maxLightItemSet', () => {
```
Comment (bhollis) [inline src/app/loadout-drawer/auto-loadouts.test.ts:45]: IMO these tests are too specific and generally low value

**Review by bhollis — COMMENTED** [review]
This is great. I'd tried to have Claude do this maybe 6 months ago and it couldn't really hack it.

File: src/testing/move-item-test-utils.ts (hunk 2 above)
Comment (delphiactual) [reply src/testing/move-item-test-utils.ts:50]: Good call, no reason to. I reworked all the fixture helpers to be pure and immutable. They take the stores array and return a new one (`addItemToStore` and `placeItemInPostmaster` return `[stores, placedItem]`, the rest return new stores). Touched stores get a fresh identity, so the `findItemsByBucket` weak-memo stays valid, which let me delete `itemsInBucketUncached` entirely.

File: src/testing/move-item-test-utils.ts (hunk 1 above)
Comment (delphiactual) [reply src/testing/move-item-test-utils.ts]: Done. With the helpers immutable, `setupMoveTestStore` no longer needs to refresh store identities before dispatching, so that hack is gone too. I used plain spread updates rather than immer since the helpers are small, but I'm happy to switch to `produce` if you'd prefer the consistency.

File: src/app/inventory/item-move-service.ts (hunk 4 above)
Comment (delphiactual) [reply src/app/inventory/item-move-service.ts]: Agreed. I collapsed them into a single `involvedItems: Exclusion[]` carrying both id and hash. The consumables-penalty check now does a hash `.some()` over it, and it's passed straight to `getSimilarItem` as the exclusion list.

File: src/app/inventory/item-move-service.ts (hunk 5 above)
Comment (delphiactual) [reply src/app/inventory/item-move-service.ts]: I documented the precondition (callers should have the items on the store, and `loadout-apply` does) and kept the loop as a safety net specifically for the `getSimilarItem` dequip replacements, which can still come from the vault. If you'd rather I move those onto the store at the `loadout-apply` callsite and drop the guard here, I'm happy to. Let me know which you prefer.

File: src/app/inventory/item-move-service.test.ts (hunk 6 above)
Comment (delphiactual) [reply src/app/inventory/item-move-service.test.ts]: Done. All four find helpers now throw a descriptive error when nothing matches, and I removed the redundant `expect(...).toBeDefined()` at the callsites.

File: src/app/inventory/item-move-service.test.ts (hunk 7 above)
Comment (delphiactual) [reply src/app/inventory/item-move-service.test.ts]: Nope. Moved `buildFreshStores()` into `beforeEach`.

File: src/app/inventory/item-move-service.test.ts (hunk 12 above)
Comment (delphiactual) [reply src/app/inventory/item-move-service.test.ts:296]: This pushed me to add a parameterized matrix in a new file, `item-move-matrix.test.ts`, using `describe.each` tables per axis:

- source by destination (vault, current, other) for an instanced item
- destination fullness: room, bucket-full-but-moveable, full-and-unmovable, asserting success or no-space (262)
- stack merge and overflow at both vault and character destinations: dest has none, a partial stack with room, or a stack that overflows into a second, asserting both conservation and the physical-stack count (543), plus the unique-stack single-item refusal
- postmaster pulls across destination and fullness states (313)

The stack rows use vault-to-char since a char-to-char consumable move just redirects to the current character. I also left out the "vault full and buckets full but another character has room, so it cascades" combo (262) for now, since it's hard to set up deterministically without the API rejecting move-asides. I took it as a per-axis table rather than a single auto-generated cartesian. Let me know if you'd like it expanded toward the fully-generated version.

File: src/app/loadout-drawer/auto-loadouts.test.ts (hunk 13 above)
Comment (delphiactual) [reply src/app/loadout-drawer/auto-loadouts.test.ts:45]: Trimmed to just the #11648 ghost-shell regression, and dropped the artifact, highest-light, and D2-ghost cases.

File: src/app/loadout-drawer/loadout-apply.ts (hunk 3 above)
Comment (delphiactual) [reply src/app/loadout-drawer/loadout-apply.ts]: comes back clean, for me

File: src/app/inventory/item-move-service.ts (hunk 5 above)
Comment (delphiactual) [reply src/app/inventory/item-move-service.ts]: Done, went with your suggestion. `equipItems` no longer moves anything itself; it now documents an on-store precondition. The dequip path in loadout-apply moves the `getSimilarItem` replacements onto the target store before calling it (a failed move there marks the dequip failed via the existing catch). Pushed in d615d99.

File: src/app/inventory/item-move-service.test.ts (hunk 8 above)
Comment (delphiactual) [reply src/app/inventory/item-move-service.test.ts]: Added these to the matrix (`item-move-matrix.test.ts`). There's a vault-fill helper (`setVaultBucketFull`), a "vault full but another character has room" case that asserts the move-aside cascades onto another character, and an "everything full" case that asserts a clean no-space failure. I didn't assert the movable-but-deep-cascade failure. With a reserved slot on the destination the engine can shuffle bumps around, so it isn't deterministic. The cascade success case covers the positive path.

File: src/app/inventory/item-move-service.test.ts (hunk 9 above)
Comment (delphiactual) [reply src/app/inventory/item-move-service.test.ts]: Extended the postmaster table with the other-character-bucket-full combo, and added a pull-into-the-vault case, alongside the existing own-character room/full and other-character-room rows.

File: src/app/inventory/item-move-service.test.ts (hunk 10 above)
Comment (delphiactual) [reply src/app/inventory/item-move-service.test.ts]: The source-by-destination matrix runs for an instanced weapon. Stacks are covered by the dedicated merge/overflow blocks (vault and current destinations) plus two new cases: a character-to-vault stack move, and the account-wide redirect when a stack is sent to a non-current character (it resolves onto the current one).

File: src/app/inventory/item-move-service.test.ts (hunk 11 above)
Comment (delphiactual) [reply src/app/inventory/item-move-service.test.ts:265]: These are already covered by the `stackCases` table (destination has none, a partial stack with room, or a stack that overflows into a second) plus the "refuses to move a unique stack onto a destination already at max" test for the single-item no-overflow case.

File: src/app/inventory/item-move-matrix.test.ts [hunk 14]
```diff
@@ -0,0 +1,493 @@
...
+  equipItems: jest.fn(),
+  setLockState: jest.fn().mockResolvedValue({}),
+  setTrackedState: jest.fn().mockResolvedValue({}),
+}));
+
+const transferMock = transfer as jest.Mock;
+const equipMock = equip as jest.Mock;
+
+/**
+ * A parameterized "movement matrix" for the item-move-service - the
```
Comment (bhollis) [inline src/app/inventory/item-move-matrix.test.ts:37]: This is neat, though I had expected it to replace the individual tests in `item-move-service.test.ts`.

**Review by bhollis — APPROVED** [review]
I'm excited about this!
